packages/frontend/src/providers/TrackingProvider.tsx

```diff
@@ -0,0 +1,217 @@
+            initial_referrer: null,
+            search: null,
+        }),
+        [],
+    );
+
+    const lightdashContext = useMemo(
+        () => ({
+            app: appContext,
+            page: getLightdashPageProperties(pageContext),
+        }),
+        [appContext, pageContext, getLightdashPageProperties],
+    );
+
+    useEffect(() => {
+        if (rudder) {
+            setRudderAnalytics(rudder);
+        } else if (writeKey && dataPlaneUrl) {
+            rudderSDK.load(writeKey, dataPlaneUrl);
+            rudderSDK.ready(() => {
+                setRudderAnalytics(rudderSDK);
+            });
+        }
+    }, [rudder, writeKey, dataPlaneUrl]);
+
+    const page = useCallback(
+        (rudderPageEvent: PageData): void => {
+            console.log(
+                'page',
+                rudderPageEvent.category,
+                rudderPageEvent.name,
+                getLightdashPageProperties(rudderPageEvent),
+                lightdashContext,
+            );
+            rudderAnalytics?.page(
+                rudderPageEvent.category,
+                rudderPageEvent.name,
+                getLightdashPageProperties(rudderPageEvent),
+                lightdashContext,
+            );
+        },
+        [rudderAnalytics, lightdashContext, getLightdashPageProperties],
+    );
+
+    const track = useCallback(
+        ({ name, properties = {} }: EventData): void => {
+            console.log('track', `${LIGHTDASH_APP_NAME}.${name}`, properties, {
+                ...lightdashContext,
+                section: sectionContext,
+            });
+            rudderAnalytics?.track(
+                `${LIGHTDASH_APP_NAME}.${name}`,
+                properties,
+                {
+                    ...lightdashContext,
+                    section: sectionContext,
+                },
+            );
+        },
+        [rudderAnalytics, sectionContext, lightdashContext],
+    );
+    const identify = useCallback(
+        ({ id, traits }: IdentifyData) => {
+            console.log('identify', id, traits, lightdashContext);
+            if (mode && mode !== LightdashMode.DEMO) {
+                rudderAnalytics?.identify(id, traits, lightdashContext);
+            }
+        },
+        [lightdashContext, mode, rudderAnalytics],
+    );
+
+    const context = useMemo(
+        () => ({
+            data: { rudder, page: pageContext, section: sectionContext },
+            page,
+            track,
+            identify,
+        }),
+        [rudder, pageContext, sectionContext, page, track, identify],
+    );
+    return (
+        <Context.Provider value={context}>{children || null}</Context.Provider>
+    );
+};
+
+export function useTracking(): TrackingContext {
+    const context = useContext(Context);
+    if (context === undefined) {
+        throw new Error('useTracking must be used within a TrackingProvider');
+    }
+    return context;
+}
+
+const NestedTrackingProvider: FC<Partial<TrackingData>> = ({
```

Comment:
Why is it called `NestedTrackingProvider`? What's the nested bit?

Reply:
The purpose of this component is to create a new tracking context by extending its parent tracking context. 

```
<TrackingProvider> // creates main context
    <Page name="myPage" > // creates page context from main context 
        <Section name="mySection"/> // creates section context from page context 
    </Page>
</TrackingProvider>
```

This component can only be used inside a TrackingProvider, that's why is called `NestedTrackingProvider`.

Reply:
Oh nice!!